### ISSUE TITLE
[AI] Active Learning Pipeline for Continuous Classifier Improvement

### ISSUE THREAD

**Issue by Daksh7785** [riteshbonthalakoti/HELPDESK.AI#1933]
<html>
<body>
<html><head></head><body><h1>🚀 [AI] Active Learning Pipeline for Continuous Classifier Improvement</h1><h2>Overview</h2><p>HelpDesk.AI currently relies on a DistilBERT-based ticket classification model that achieves approximately <strong>85% accuracy</strong> during offline evaluation. However, the classifier operates as a static model and lacks mechanisms to learn from real-world usage after deployment.</p><p>While administrator corrections are already recorded in <code inline="">corrections_log.json</code>, the system does not leverage this valuable feedback for retraining, model evaluation, or continuous improvement. Over time, this results in model drift, declining classification quality, and increased manual intervention.</p><p>This feature proposes implementing a complete <strong>Active Learning + Continuous Retraining (MLOps) Pipeline</strong> that transforms administrator feedback into actionable training data, enabling the classifier to improve automatically over time.</p><hr><h1>Problem Statement</h1><p>The current machine learning workflow follows a traditional one-time training approach:</p><pre><code class="language-text">Train Model → Deploy → Predict
</code></pre><p>There is no feedback loop between production usage and future model versions.</p><h3>Key Limitations</h3><h3>1. Data Drift</h3><p>The production environment continuously evolves:</p><ul><li><p>New ticket categories emerge</p></li><li><p>Product terminology changes</p></li><li><p>Customer vocabulary differs from training data</p></li><li><p>New integrations introduce unseen issues</p></li></ul><p>As a result, prediction accuracy gradually deteriorates.</p><hr><h3>2. No Continuous Learning</h3><p>Although administrators frequently correct ticket classifications, the model never learns from these corrections.</p><p>Example:</p><pre><code class="language-json">{
  "ticket": "VPN authentication failed",
  "prediction": "Software Issue",
  "correct_label": "Network Issue"
}
</code></pre><p>This correction represents high-quality labeled data but is currently discarded after logging.</p><hr><h3>3. Static Model Lifecycle</h3><p>Current workflow:</p><pre><code class="language-text">Training Dataset
       ↓
Train DistilBERT
       ↓
Deploy
       ↓
Never Updated Again
</code></pre><p>This prevents adaptation to evolving support environments.</p><hr><h3>4. Unused Ground Truth Data</h3><p>The existing:</p><pre><code class="language-bash">corrections_log.json
</code></pre><p>contains verified labels generated by administrators and support teams.</p><p>These corrections are among the most valuable sources of training data because they originate directly from production traffic.</p><hr><h3>5. No Model Governance</h3><p>The platform currently lacks:</p><ul><li><p>Model version tracking</p></li><li><p>Performance benchmarking</p></li><li><p>Safe deployment mechanisms</p></li><li><p>Rollback capabilities</p></li><li><p>Retraining automation</p></li></ul><hr><h1>Why This Matters</h1><h2>Business Impact</h2><p>Classification quality directly affects:</p><ul><li><p>Ticket routing accuracy</p></li><li><p>Resolution times</p></li><li><p>SLA compliance</p></li><li><p>Support efficiency</p></li><li><p>Customer satisfaction</p></li></ul><p>Even minor degradation can significantly increase operational workload.</p><hr><h2>Leveraging Human Expertise</h2><p>Every administrator correction represents domain expertise.</p><p>Without a feedback pipeline:</p><pre><code class="language-text">Human Knowledge → Lost
</code></pre><p>With active learning:</p><pre><code class="language-text">Human Knowledge
      ↓
Training Data
      ↓
Model Improvement
</code></pre><hr><h2>Competitive Positioning</h2><p>Modern support platforms continuously improve through:</p><ul><li><p>Human feedback loops</p></li><li><p>Active learning systems</p></li><li><p>Automated retraining</p></li><li><p>Continuous model evaluation</p></li></ul><p>Implementing these capabilities improves HelpDesk.AI's long-term competitiveness.</p><hr><h1>Real-World Impact</h1><h2>Scenario 1: Product Evolution</h2><p>A company launches a new feature.</p><p>Customers begin submitting tickets using previously unseen terminology.</p><h3>Current Result</h3><p>Classifier confidence decreases and routing quality suffers.</p><h3>Desired Result</h3><p>Administrator corrections become training examples and the model gradually learns new terminology.</p><hr><h2>Scenario 2: Repeated Misclassifications</h2><p>The classifier repeatedly confuses:</p><pre><code class="language-text">Network Issues
↔
Authentication Issues
</code></pre><p>Administrators correct the same mistake hundreds of times.</p><h3>Current State</h3><p>Every correction is isolated.</p><h3>Desired State</h3><p>Corrections become part of future training datasets, eliminating recurring mistakes.</p><hr><h2>Scenario 3: Enterprise Customers</h2><p>Healthcare organizations may submit tickets containing:</p><ul><li><p>EMR</p></li><li><p>HL7</p></li><li><p>PACS</p></li><li><p>HIPAA</p></li></ul><p>Financial organizations may submit tickets containing:</p><ul><li><p>SWIFT</p></li><li><p>ACH</p></li><li><p>Payment Gateway</p></li><li><p>Settlement Systems</p></li></ul><p>These terms may not exist in the original training dataset.</p><p>Continuous learning allows adaptation to customer-specific domains.</p><hr><h1>Proposed Solution</h1><p>Implement a comprehensive Active Learning Pipeline consisting of six major components.</p><hr><h1>Phase 1: Feedback Collection</h1><h2>Enhanced Correction Logging</h2><p>Extend existing correction tracking with additional metadata.</p><h3>Current Structure</h3><pre><code class="language-json">{
  "ticket_text": "...",
  "predicted_label": "...",
  "corrected_label": "..."
}
</code></pre><h3>Proposed Structure</h3><pre><code class="language-json">{
  "ticket_id": "123",
  "ticket_text": "...",
  "predicted_label": "...",
  "corrected_label": "...",
  "prediction_confidence": 0.67,
  "classifier_version": "v1.3.2",
  "tenant_id": "tenant_abc",
  "corrected_by": "admin_id",
  "created_at": "timestamp"
}
</code></pre><h3>Additional Telemetry</h3><p>Track:</p><ul><li><p>Prediction confidence</p></li><li><p>Category frequency</p></li><li><p>Misclassification trends</p></li><li><p>Class imbalance statistics</p></li><li><p>Model version metrics</p></li></ul><hr><h1>Phase 2: Training Data Preparation</h1><h2>Automated Weekly Data Extraction</h2><p>Create a scheduled job that:</p><ol><li><p>Reads corrections from previous 7 days</p></li><li><p>Validates labels</p></li><li><p>Removes duplicates</p></li><li><p>Resolves conflicting annotations</p></li><li><p>Creates retraining dataset</p></li></ol><hr><h2>Hard Negative Mining</h2><p>Identify examples where the model was:</p><pre><code class="language-text">Highly Confident
AND
Incorrect
</code></pre><p>Example:</p><pre><code class="language-text">Prediction: Billing
Confidence: 97%
Actual: Technical Issue
</code></pre><p>These samples provide exceptionally valuable learning signals.</p><hr><h2>Dataset Balancing</h2><p>Prevent dominant classes from overwhelming retraining.</p><p>Techniques may include:</p><ul><li><p>Class weighting</p></li><li><p>Oversampling minority classes</p></li><li><p>Stratified sampling</p></li></ul><hr><h2>Noise Reduction</h2><p>Detect and remove:</p><ul><li><p>Duplicate corrections</p></li><li><p>Invalid labels</p></li><li><p>Contradictory annotations</p></li><li><p>Outlier samples</p></li></ul><hr><h1>Phase 3: Automated Retraining Pipeline</h1><h2>Scheduled Retraining</h2><p>Frequency:</p><pre><code class="language-text">Every 14 Days
</code></pre><p>GitHub Workflow:</p><pre><code class="language-bash">.github/workflows/retrain-classifier.yml
</code></pre><p>Pipeline:</p><pre><code class="language-text">Collect Corrections
       ↓
Prepare Dataset
       ↓
Fine-Tune DistilBERT
       ↓
Evaluate
       ↓
Candidate Model
</code></pre><hr><h2>Fine-Tuning Strategy</h2><p>Retrain using:</p><ul><li><p>Existing training corpus</p></li><li><p>Recent correction data</p></li><li><p>Hard negative samples</p></li></ul><p>Preserve previously learned knowledge while adapting to recent patterns.</p><hr><h1>Phase 4: Model Evaluation &amp; Promotion</h1><p>Every candidate model must pass validation before deployment.</p><h2>Evaluation Metrics</h2><p>Track:</p>
Metric | Purpose
-- | --
Accuracy | Overall correctness
Precision | False positive control
Recall | Missed classifications
F1 Score | Balanced performance
Confidence Calibration | Prediction reliability

<hr><h1>Acceptance Criteria</h1><h2>Feedback Collection</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Correction logs include confidence scores and model version metadata</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Prediction telemetry captured successfully</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Misclassification statistics tracked</p></li></ul><h2>Data Preparation</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Weekly correction extraction job implemented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Hard negatives identified automatically</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Dataset balancing prevents class skew</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Noisy annotations filtered before training</p></li></ul><h2>Retraining</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Bi-weekly retraining workflow executes automatically</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> DistilBERT fine-tuning pipeline implemented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Validation metrics generated for every training run</p></li></ul><h2>Evaluation</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Candidate models evaluated against production baseline</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Shadow evaluation runs on production traffic</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Deployment blocked when improvement &lt; 2%</p></li></ul><h2>Governance</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Version history maintained</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Rollback mechanism implemented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Deployment audit logs retained</p></li></ul><h2>Active Learning</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Low-confidence predictions identified</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Hard negatives prioritized</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Feedback incorporated into future training cycles</p></li></ul><h2>Enterprise Support</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Tenant-specific models supported</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Automatic fallback to global model available</p></li></ul><h2>Documentation</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Active learning architecture documented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Retraining workflow documented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Model deployment strategy documented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Rollback procedures documented</p></li></ul><hr><h1>Issue Metadata</h1><p><strong>Difficulty:</strong> Advanced</p><p><strong>Labels:</strong> <code inline="">ai</code>, <code inline="">machine-learning</code>, <code inline="">mlops</code>, <code inline="">backend</code>, <code inline="">automation</code></p><p><strong>Impact Score:</strong> 9/10</p><p><strong>Effort Score:</strong> 8/10</p><p><strong>Priority:</strong> 🔴 HIGH</p><hr><h2>Additional Notes</h2><p>This feature establishes the foundation for a fully managed machine learning lifecycle within HelpDesk.AI. By integrating administrator feedback, automated retraining, shadow deployments, model governance, and active learning strategies, the classifier can continuously improve while maintaining production stability, explainability, and enterprise-grade operational controls.</p></body></html>
</body>
</html>

**Comment by riteshbonthalakoti** [comment]
Hi GSSoC contributors! 🙌

Thank you so much for your interest in contributing to HELPDESK.AI! This issue is open and available for ALL contributors! 🚀

Under GSSoC rules, multiple contributors can work on the same issue and the best, most comprehensive work will get merged. 

If you want to tackle this, please complete the mandatory onboarding steps below and start coding! 🚀💻


### 🌟 Mandatory Onboarding Steps (Must Complete!)
To ensure your GSSoC contribution is evaluated and your account cleared for dashboard testing, please complete these steps:
1. ⭐ **Star this repository**: https://github.com/ritesh-1918/HELPDESK.AI
2. 🍴 **Fork this repository**: https://github.com/ritesh-1918/HELPDESK.AI/fork
3. 👤 **Follow @ritesh-1918 on GitHub**: https://github.com/ritesh-1918
4. 💼 **Connect on LinkedIn**: https://www.linkedin.com/in/ritesh1908/
5. 🚀 **Register & Onboard**:
   - Go to the deployed app: https://helpdeskaiv1.vercel.app/
   - Click **Sign In** -> **Create Account**
   - Select **Ritesh Private Limited Company** as your organization.
   - Reply to this thread or email ritesh at bonthalamadhavi1@gmail.com with your username to get your access approved instantly!

*Note: All PR branches must target the `gssoc` branch, NOT `main`.*

**Comment by riteshbonthalakoti** [comment]
Resolved by merging all relevant GSSoC contributions. Closing issue. Thank you!
